File: zig/lispier/src/ast.zig
const std = @import("std");
const reader_types = @import("reader_types.zig");

/// AST Node types
pub const NodeType = enum {
    Module,
    Operation,
    Region,
    Block,
    Def,
    Let,
    TypeAnnotation,
    FunctionType,
    Literal,
};

/// Type representation
pub const Type = struct {
    name: []const u8,
    allocator: std.mem.Allocator,

    pub fn init(allocator: std.mem.Allocator, name: []const u8) !*Type {
        const t = try allocator.create(Type);
        t.* = .{
            .name = try allocator.dupe(u8, name),
            .allocator = allocator,
        };
        return t;
    }

    pub fn deinit(self: *Type) void {
        self.allocator.free(self.name);
        self.allocator.destroy(self);
    }
};

/// Function type (-> [args] [returns])
pub const FunctionType = struct {
    arg_types: std.ArrayList(*Type),
    return_types: std.ArrayList(*Type),
    allocator: std.mem.Allocator,

    pub fn init(allocator: std.mem.Allocator) *FunctionType {
        const ft = allocator.create(FunctionType) catch unreachable;
        ft.* = .{
            .arg_types = std.ArrayList(*Type){},
            .return_types = std.ArrayList(*Type){},
            .allocator = allocator,
        };
        return ft;
    }

    pub fn deinit(self: *FunctionType) void {
        for (self.arg_types.items) |t| {
            t.deinit();
        }
        self.arg_types.deinit(self.allocator);

        for (self.return_types.items) |t| {
            t.deinit();
        }
        self.return_types.deinit(self.allocator);

        self.allocator.destroy(self);
    }
};

/// Typed number (number with explicit type)
pub const TypedNumber = struct {
    value: f64,
    typ: *Type,
};

/// Attribute value (for operation attributes)
pub const AttributeValue = union(enum) {
    string: []const u8,
    number: f64,
    boolean: bool,
    array: std.ArrayList(AttributeValue),
    type: *Type,
    function_type: *FunctionType,
    typed_number: TypedNumber,
};

/// Module (top-level container)
pub const Module = struct {
    body: std.ArrayList(*Node),
    allocator: std.mem.Allocator,

    pub fn init(allocator: std.mem.Allocator) !*Module {
        const module = try allocator.create(Module);
        module.* = .{
            .body = std.ArrayList(*Node){},
            .allocator = allocator,
        };
        return module;
    }

    pub fn deinit(self: *Module) void {
        for (self.body.items) |n| {
            n.deinit();
        }
        self.body.deinit(self.allocator);
        self.allocator.destroy(self);
    }
};

/// Operation (MLIR operation)
pub const Operation = struct {
    name: []const u8,
    namespace: ?[]const u8,
    attributes: std.StringHashMap(AttributeValue),
    operands: std.ArrayList(*Node),
    regions: std.ArrayList(*Region),
    result_types: std.ArrayList(*Type),
    allocator: std.mem.Allocator,

    pub fn init(allocator: std.mem.Allocator, name: []const u8, namespace: ?[]const u8) !*Operation {
        const op = try allocator.create(Operation);
        op.* = .{
            .name = try allocator.dupe(u8, name),
            .namespace = if (namespace) |ns| try allocator.dupe(u8, ns) else null,
            .attributes = std.StringHashMap(AttributeValue).init(allocator),
            .operands = std.ArrayList(*Node){},
            .regions = std.ArrayList(*Region){},
            .result_types = std.ArrayList(*Type){},
            .allocator = allocator,
        };
        return op;
    }

    pub fn deinit(self: *Operation) void {
        self.allocator.free(self.name);
        if (self.namespace) |ns| {
            self.allocator.free(ns);
        }

        var it = self.attributes.iterator();
        while (it.next()) |entry| {
            self.allocator.free(entry.key_ptr.*);
            deinitAttributeValue(entry.value_ptr.*, self.allocator);
        }
        self.attributes.deinit();

        for (self.operands.items) |operand| {
            operand.deinit();
        }
        self.operands.deinit(self.allocator);

        for (self.regions.items) |region| {
            region.deinit();
        }
        self.regions.deinit(self.allocator);

        for (self.result_types.items) |t| {
            t.deinit();
        }
        self.result_types.deinit(self.allocator);

        self.allocator.destroy(self);
    }

    fn deinitAttributeValue(value_: AttributeValue, allocator: std.mem.Allocator) void {
        var value = value_;
        switch (value) {
            .string => |s| allocator.free(s),
            .array => |*arr| {
                for (arr.items) |item| {
                    deinitAttributeValue(item, allocator);
                }
                arr.deinit(allocator);
            },
            .type => |t| t.deinit(),
            .function_type => |ft| ft.deinit(),
            .typed_number => |tn| tn.typ.deinit(),
            else => {},
        }
    }

    pub fn getQualifiedName(self: *const Operation, allocator: std.mem.Allocator) ![]const u8 {
        if (self.namespace) |ns| {
            return std.fmt.allocPrint(allocator, "{s}.{s}", .{ ns, self.name });
        }
        return allocator.dupe(u8, self.name);
    }
};

/// Region (contains blocks)
pub const Region = struct {
    blocks: std.ArrayList(*Block),
    allocator: std.mem.Allocator,

    pub fn init(allocator: std.mem.Allocator) !*Region {
        const region = try allocator.create(Region);
        region.* = .{
            .blocks = std.ArrayList(*Block){},
            .allocator = allocator,
        };
        return region;
    }

    pub fn deinit(self: *Region) void {
        for (self.blocks.items) |block| {
            block.deinit();
        }
        self.blocks.deinit(self.allocator);
        self.allocator.destroy(self);
    }
};

/// Block argument
pub const BlockArgument = struct {
    name: []const u8,
    type: ?*Type,
    allocator: std.mem.Allocator,

    pub fn init(allocator: std.mem.Allocator, name: []const u8, arg_type: ?*Type) !*BlockArgument {
        const arg = try allocator.create(BlockArgument);
        arg.* = .{
            .name = try allocator.dupe(u8, name),
            .type = arg_type,
            .allocator = allocator,
        };
        return arg;
    }

    pub fn deinit(self: *BlockArgument) void {
        self.allocator.free(self.name);
        if (self.type) |t| {
            t.deinit();
        }
        self.allocator.destroy(self);
    }
};

/// Block (labeled block with arguments)
pub const Block = struct {
    label: ?[]const u8,
    arguments: std.ArrayList(*BlockArgument),
    operations: std.ArrayList(*Node),
    allocator: std.mem.Allocator,

    pub fn init(allocator: std.mem.Allocator, label: ?[]const u8) !*Block {
        const block = try allocator.create(Block);
        block.* = .{
            .label = if (label) |l| try allocator.dupe(u8, l) else null,
            .arguments = std.ArrayList(*BlockArgument){},
            .operations = std.ArrayList(*Node){},
            .allocator = allocator,
        };
        return block;
    }

    pub fn deinit(self: *Block) void {
        if (self.label) |l| {
            self.allocator.free(l);
        }

        for (self.arguments.items) |arg| {
            arg.deinit();
        }
        self.arguments.deinit(self.allocator);

        for (self.operations.items) |op| {
            op.deinit();
        }
        self.operations.deinit(self.allocator);

        self.allocator.destroy(self);
    }
};

/// Binding (def or let)
pub const Binding = struct {
    names: std.ArrayList([]const u8),
    value: *Node,
    allocator: std.mem.Allocator,

    pub fn init(allocator: std.mem.Allocator) !*Binding {
        const binding = try allocator.create(Binding);
        binding.* = .{
            .names = std.ArrayList([]const u8){},
            .value = undefined, // Must be set after creation
            .allocator = allocator,
        };
        return binding;
    }

    pub fn deinit(self: *Binding) void {
        for (self.names.items) |name| {
            self.allocator.free(name);
        }
        self.names.deinit(self.allocator);
        self.value.deinit();
        self.allocator.destroy(self);
    }
};

/// Let expression: bindings vector and body expressions
pub const LetExpr = struct {
    bindings: std.ArrayList(*Binding),
    body: std.ArrayList(*Node),
    allocator: std.mem.Allocator,

    pub fn init(allocator: std.mem.Allocator) !*LetExpr {
        const le = try allocator.create(LetExpr);
        le.* = .{
            .bindings = std.ArrayList(*Binding){},
            .body = std.ArrayList(*Node){},
            .allocator = allocator,
        };
        return le;
    }

    pub fn deinit(self: *LetExpr) void {
        for (self.bindings.items) |b| {
            b.deinit();
        }
        self.bindings.deinit(self.allocator);

        for (self.body.items) |n| {
            n.deinit();
        }
        self.body.deinit(self.allocator);

        self.allocator.destroy(self);
    }
};

/// Literal value
pub const Literal = struct {
    value: reader_types.Value,

    pub fn deinit(self: *Literal, allocator: std.mem.Allocator) void {
        // Note: Value owns its own memory
        _ = allocator;
        _ = self;
    }
};

/// AST Node
pub const Node = struct {
    node_type: NodeType,
    data: union {
        module: *Module,
        operation: *Operation,
        region: *Region,
        block: *Block,
        binding: *Binding,
        let_expr: *LetExpr,
        type_annotation: struct {
            value: *Node,
            typ: *Type,
        },
        function_type: *FunctionType,
        literal: *reader_types.Value,
    },
    allocator: std.mem.Allocator,

    pub fn createModule(allocator: std.mem.Allocator, module: *Module) !*Node {
        const node = try allocator.create(Node);
        node.* = .{
            .node_type = .Module,
            .data = .{ .module = module },
            .allocator = allocator,
        };
        return node;
    }

    pub fn createOperation(allocator: std.mem.Allocator, operation: *Operation) !*Node {
        const node = try allocator.create(Node);
        node.* = .{
            .node_type = .Operation,
            .data = .{ .operation = operation },
            .allocator = allocator,
        };
        return node;
    }

    pub fn createRegion(allocator: std.mem.Allocator, region: *Region) !*Node {
        const node = try allocator.create(Node);
        node.* = .{
            .node_type = .Region,
            .data = .{ .region = region },
            .allocator = allocator,
        };
        return node;
    }

    pub fn createBlock(allocator: std.mem.Allocator, block: *Block) !*Node {
        const node = try allocator.create(Node);
        node.* = .{
            .node_type = .Block,
            .data = .{ .block = block },
            .allocator = allocator,
        };
        return node;
    }

    pub fn createBinding(allocator: std.mem.Allocator, binding: *Binding) !*Node {
        const node = try allocator.create(Node);
        node.* = .{
            .node_type = .Def,
            .data = .{ .binding = binding },
            .allocator = allocator,
        };
        return node;
    }

    pub fn createLet(allocator: std.mem.Allocator, let_expr: *LetExpr) !*Node {
        const node = try allocator.create(Node);
        node.* = .{
            .node_type = .Let,
            .data = .{ .let_expr = let_expr },
            .allocator = allocator,
        };
        return node;
    }

    pub fn createLiteral(allocator: std.mem.Allocator, value: *reader_types.Value) !*Node {
        const node = try allocator.create(Node);
        node.* = .{
            .node_type = .Literal,
            .data = .{ .literal = value },
            .allocator = allocator,
        };
        return node;
    }

    pub fn createTypeAnnotation(allocator: std.mem.Allocator, value: *Node, typ: *Type) !*Node {
        const node = try allocator.create(Node);
        node.* = .{
            .node_type = .TypeAnnotation,
            .data = .{ .type_annotation = .{ .value = value, .typ = typ } },
            .allocator = allocator,
        };
        return node;
    }

    pub fn createFunctionType(allocator: std.mem.Allocator, ft: *FunctionType) !*Node {
        const node = try allocator.create(Node);
        node.* = .{
            .node_type = .FunctionType,
            .data = .{ .function_type = ft },
            .allocator = allocator,
        };
        return node;
    }

    pub fn deinit(self: *Node) void {
        switch (self.node_type) {
            .Module => self.data.module.deinit(),
            .Operation => self.data.operation.deinit(),
            .Region => self.data.region.deinit(),
            .Block => self.data.block.deinit(),
            .Def => self.data.binding.deinit(),
            .Let => self.data.let_expr.deinit(),
            .TypeAnnotation => {
                self.data.type_annotation.value.deinit();
                self.data.type_annotation.typ.deinit();
            },
            .FunctionType => self.data.function_type.deinit(),
            .Literal => {}, // Don't deinit - values are owned by the values list
        }
        self.allocator.destroy(self);
    }
};

File: claude-experiments/lispier/src/ast.rs
use std::collections::HashMap;

use crate::value::Value;

/// Type representation
#[derive(Debug, Clone, PartialEq)]
pub struct Type {
    pub name: String,
}

impl Type {
    pub fn new(name: impl Into<String>) -> Self {
        Self { name: name.into() }
    }
}

/// Function type (-> [args] [returns])
#[derive(Debug, Clone, PartialEq)]
pub struct FunctionType {
    pub arg_types: Vec<Type>,
    pub return_types: Vec<Type>,
}

impl FunctionType {
    pub fn new() -> Self {
        Self {
            arg_types: Vec::new(),
            return_types: Vec::new(),
        }
    }
}

impl Default for FunctionType {
    fn default() -> Self {
        Self::new()
    }
}

/// Typed number (number with explicit type)
#[derive(Debug, Clone, PartialEq)]
pub struct TypedNumber {
    pub value: f64,
    pub typ: Type,
}

/// Typed MLIR literal (e.g., dense<[1,2,3]> with type tensor<3xi32>)
#[derive(Debug, Clone, PartialEq)]
pub struct TypedMLIRLiteral {
    pub literal: String,
    pub typ: Type,
}

/// Attribute value (for operation attributes)
#[derive(Debug, Clone, PartialEq)]
pub enum AttributeValue {
    String(String),
    Number(f64),
    Boolean(bool),
    Array(Vec<AttributeValue>),
    Type(Type),
    FunctionType(FunctionType),
    TypedNumber(TypedNumber),
    /// Typed MLIR attribute literal: (: dense<[1,2,3]> tensor<3xi32>)
    TypedMLIRLiteral(TypedMLIRLiteral),
    /// MLIR attribute literal syntax like array<i32: 0, 1, 1>, dense<...>, etc.
    /// Stored with spaces already converted to commas for MLIR parsing.
    MLIRLiteral(String),
}

/// Block argument
#[derive(Debug, Clone, PartialEq)]
pub struct BlockArgument {
    pub name: String,
    pub typ: Option<Type>,
}

impl BlockArgument {
    pub fn new(name: impl Into<String>) -> Self {
        Self {
            name: name.into(),
            typ: None,
        }
    }

    pub fn with_type(name: impl Into<String>, typ: Type) -> Self {
        Self {
            name: name.into(),
            typ: Some(typ),
        }
    }
}

/// Block (labeled block with arguments)
#[derive(Debug, Clone, PartialEq)]
pub struct Block {
    pub label: Option<String>,
    pub arguments: Vec<BlockArgument>,
    pub operations: Vec<Node>,
}

impl Block {
    pub fn new() -> Self {
        Self {
            label: None,
            arguments: Vec::new(),
            operations: Vec::new(),
        }
    }

    pub fn with_label(label: impl Into<String>) -> Self {
        Self {
            label: Some(label.into()),
            arguments: Vec::new(),
            operations: Vec::new(),
        }
    }
}

impl Default for Block {
    fn default() -> Self {
        Self::new()
    }
}

/// Region (contains blocks)
#[derive(Debug, Clone, PartialEq)]
pub struct Region {
    pub blocks: Vec<Block>,
}

impl Region {
    pub fn new() -> Self {
        Self { blocks: Vec::new() }
    }
}

impl Default for Region {
    fn default() -> Self {
        Self::new()
    }
}

/// Operation (MLIR operation)
#[derive(Debug, Clone, PartialEq)]
pub struct Operation {
    pub name: String,
    pub namespace: Option<String>,
    pub attributes: HashMap<String, AttributeValue>,
    pub operands: Vec<Node>,
    pub regions: Vec<Region>,
    pub result_types: Vec<Type>,
}

impl Operation {
    pub fn new(name: impl Into<String>) -> Self {
        Self {
            name: name.into(),
            namespace: None,
            attributes: HashMap::new(),
            operands: Vec::new(),
            regions: Vec::new(),
            result_types: Vec::new(),
        }
    }

    pub fn with_namespace(name: impl Into<String>, namespace: impl Into<String>) -> Self {
        Self {
            name: name.into(),
            namespace: Some(namespace.into()),
            attributes: HashMap::new(),
            operands: Vec::new(),
            regions: Vec::new(),
            result_types: Vec::new(),
        }
    }

    pub fn qualified_name(&self) -> String {
        if let Some(ref ns) = self.namespace {
            format!("{}.{}", ns, self.name)
        } else {
            self.name.clone()
        }
    }
}

/// Module (top-level container)
#[derive(Debug, Clone, PartialEq)]
pub struct Module {
    pub body: Vec<Node>,
}

/// Require specification for loading another file
#[derive(Debug, Clone, PartialEq)]
pub struct Require {
    pub path: String,
    pub alias: String,
    pub is_project_relative: bool,
}

/// A compilation pass with optional attributes
#[derive(Debug, Clone, PartialEq)]
pub struct Pass {
    pub name: String,
    pub attributes: HashMap<String, String>,
}

impl Pass {
    pub fn new(name: impl Into<String>) -> Self {
        Self {
            name: name.into(),
            attributes: HashMap::new(),
        }
    }

    pub fn with_attributes(name: impl Into<String>, attributes: HashMap<String, String>) -> Self {
        Self {
            name: name.into(),
            attributes,
        }
    }

    /// Convert to mlir-opt pass argument format
    pub fn to_pass_arg(&self, runtime_attrs: &HashMap<String, String>) -> String {
        let mut attrs = self.attributes.clone();
        // Merge runtime attributes (runtime takes precedence)
        for (k, v) in runtime_attrs {
            attrs.insert(k.clone(), v.clone());
        }

        if attrs.is_empty() {
            format!("--{}", self.name)
        } else {
            let attr_str = attrs
                .iter()
                .map(|(k, v)| format!("{}={}", k, v))
                .collect::<Vec<_>>()
                .join(",");
            format!("--{}='{}'", self.name, attr_str)
        }
    }

    /// Convert to pipeline string format for parse_pass_pipeline
    /// Format: "pass-name" or "pass-name{attr1=value1 attr2=value2}"
    pub fn to_pipeline_string(&self, runtime_attrs: &HashMap<String, String>) -> String {
        let mut attrs = self.attributes.clone();
        // Merge runtime attributes (runtime takes precedence)
        for (k, v) in runtime_attrs {
            attrs.insert(k.clone(), v.clone());
        }

        if attrs.is_empty() {
            self.name.clone()
        } else {
            let attr_str = attrs
                .iter()
                .map(|(k, v)| format!("{}={}", k, v))
                .collect::<Vec<_>>()
                .join(" ");
            format!("{}{{{}}}", self.name, attr_str)
        }
    }
}

/// A compilation target (e.g., rocm, cuda, cpu)
#[derive(Debug, Clone, PartialEq)]
pub struct Target {
    pub backend: String,
    pub passes: Vec<Pass>,
}

impl Target {
    pub fn new(backend: impl Into<String>) -> Self {
        Self {
            backend: backend.into(),
            passes: Vec::new(),
        }
    }
}

/// Compilation specification with multiple targets
#[derive(Debug, Clone, PartialEq)]
pub struct Compilation {
    pub targets: Vec<Target>,
}

impl Compilation {
    pub fn new() -> Self {
        Self {
            targets: Vec::new(),
        }
    }

    pub fn get_target(&self, backend: &str) -> Option<&Target> {
        self.targets.iter().find(|t| t.backend == backend)
    }
}

impl Default for Compilation {
    fn default() -> Self {
        Self::new()
    }
}

/// External FFI declaration
#[derive(Debug, Clone, PartialEq)]
pub struct Extern {
    /// The FFI library to load (e.g., "value-ffi" for Value manipulation)
    pub library: String,
}

impl Extern {
    pub fn new(library: impl Into<String>) -> Self {
        Self {
            library: library.into(),
        }
    }
}

/// Link a shared library for external symbols
#[derive(Debug, Clone, PartialEq)]
pub struct LinkLibrary {
    /// The library to link
    /// Can be a keyword like :c for libc, or a path string
    pub library: String,
}

impl LinkLibrary {
    pub fn new(library: impl Into<String>) -> Self {
        Self {
            library: library.into(),
        }
    }

    /// Get the platform-specific library path
    pub fn resolve_path(&self) -> String {
        match self.library.as_str() {
            ":c" | ":libc" => {
                // Platform-specific libc - use absolute paths
                #[cfg(target_os = "macos")]
                { "/usr/lib/libSystem.B.dylib".to_string() }
                #[cfg(target_os = "linux")]
                { "/lib/x86_64-linux-gnu/libc.so.6".to_string() }
                #[cfg(not(any(target_os = "macos", target_os = "linux")))]
                { "c".to_string() }
            }
            ":m" | ":libm" => {
                #[cfg(target_os = "macos")]
                { "/usr/lib/libSystem.B.dylib".to_string() }
                #[cfg(target_os = "linux")]
                { "/lib/x86_64-linux-gnu/libm.so.6".to_string() }
                #[cfg(not(any(target_os = "macos", target_os = "linux")))]
                { "m".to_string() }
            }
            // Otherwise treat as a path or library name
            _ => self.library.clone(),
        }
    }
}

/// Macro registration declaration
/// Registers a compiled function as a macro for use during expansion
#[derive(Debug, Clone, PartialEq)]
pub struct Defmacro {
    /// The name of the function to register as a macro
    pub name: String,
}

impl Defmacro {
    pub fn new(name: impl Into<String>) -> Self {
        Self { name: name.into() }
    }
}

/// Require macros declaration
/// Loads and compiles a macro module, registering its macros for use in this module
#[derive(Debug, Clone, PartialEq)]
pub struct RequireMacros {
    /// The path to the macro module
    pub path: String,
    /// Whether the path is project-relative (starts with @)
    pub is_project_relative: bool,
}

impl RequireMacros {
    pub fn new(path: impl Into<String>) -> Self {
        let path = path.into();
        let is_project_relative = path.starts_with('@');
        Self {
            path,
            is_project_relative,
        }
    }
}

impl Module {
    pub fn new() -> Self {
        Self { body: Vec::new() }
    }
}

impl Default for Module {
    fn default() -> Self {
        Self::new()
    }
}

/// Binding (def or let)
#[derive(Debug, Clone, PartialEq)]
pub struct Binding {
    pub names: Vec<String>,
    pub value: Box<Node>,
}

impl Binding {
    pub fn new(names: Vec<String>, value: Node) -> Self {
        Self {
            names,
            value: Box::new(value),
        }
    }

    pub fn single(name: impl Into<String>, value: Node) -> Self {
        Self {
            names: vec![name.into()],
            value: Box::new(value),
        }
    }
}

/// Let expression: bindings vector and body expressions
#[derive(Debug, Clone, PartialEq)]
pub struct LetExpr {
    pub bindings: Vec<Binding>,
    pub body: Vec<Node>,
}

impl LetExpr {
    pub fn new() -> Self {
        Self {
            bindings: Vec::new(),
            body: Vec::new(),
        }
    }
}

impl Default for LetExpr {
    fn default() -> Self {
        Self::new()
    }
}

/// Type annotation
#[derive(Debug, Clone, PartialEq)]
pub struct TypeAnnotation {
    pub value: Box<Node>,
    pub typ: Type,
}

impl TypeAnnotation {
    pub fn new(value: Node, typ: Type) -> Self {
        Self {
            value: Box::new(value),
            typ,
        }
    }
}

/// AST Node
#[derive(Debug, Clone, PartialEq)]
pub enum Node {
    Module(Module),
    Operation(Operation),
    Region(Region),
    Block(Block),
    Def(Binding),
    Let(LetExpr),
    TypeAnnotation(TypeAnnotation),
    FunctionType(FunctionType),
    Literal(Value),
    Require(Require),
    RequireMacros(RequireMacros),
    Compilation(Compilation),
    Extern(Extern),
    Defmacro(Defmacro),
    LinkLibrary(LinkLibrary),
}

impl Node {
    pub fn module(module: Module) -> Self {
        Node::Module(module)
    }

    pub fn operation(operation: Operation) -> Self {
        Node::Operation(operation)
    }

    pub fn region(region: Region) -> Self {
        Node::Region(region)
    }

    pub fn block(block: Block) -> Self {
        Node::Block(block)
    }

    pub fn def(binding: Binding) -> Self {
        Node::Def(binding)
    }

    pub fn let_expr(let_expr: LetExpr) -> Self {
        Node::Let(let_expr)
    }

    pub fn type_annotation(annotation: TypeAnnotation) -> Self {
        Node::TypeAnnotation(annotation)
    }

    pub fn function_type(ft: FunctionType) -> Self {
        Node::FunctionType(ft)
    }

    pub fn literal(value: Value) -> Self {
        Node::Literal(value)
    }

    pub fn require(require: Require) -> Self {
        Node::Require(require)
    }

    pub fn require_macros(require_macros: RequireMacros) -> Self {
        Node::RequireMacros(require_macros)
    }

    pub fn compilation(compilation: Compilation) -> Self {
        Node::Compilation(compilation)
    }

    pub fn extern_decl(extern_decl: Extern) -> Self {
        Node::Extern(extern_decl)
    }

    pub fn defmacro(defmacro: Defmacro) -> Self {
        Node::Defmacro(defmacro)
    }

    pub fn is_module(&self) -> bool {
        matches!(self, Node::Module(_))
    }

    pub fn is_operation(&self) -> bool {
        matches!(self, Node::Operation(_))
    }

    pub fn is_region(&self) -> bool {
        matches!(self, Node::Region(_))
    }

    pub fn is_block(&self) -> bool {
        matches!(self, Node::Block(_))
    }

    pub fn is_def(&self) -> bool {
        matches!(self, Node::Def(_))
    }

    pub fn is_let(&self) -> bool {
        matches!(self, Node::Let(_))
    }

    pub fn is_type_annotation(&self) -> bool {
        matches!(self, Node::TypeAnnotation(_))
    }

    pub fn is_function_type(&self) -> bool {
        matches!(self, Node::FunctionType(_))
    }

    pub fn is_literal(&self) -> bool {
        matches!(self, Node::Literal(_))
    }

    pub fn is_require(&self) -> bool {
        matches!(self, Node::Require(_))
    }

    pub fn is_require_macros(&self) -> bool {
        matches!(self, Node::RequireMacros(_))
    }

    pub fn is_compilation(&self) -> bool {
        matches!(self, Node::Compilation(_))
    }

    pub fn is_extern(&self) -> bool {
        matches!(self, Node::Extern(_))
    }

    pub fn is_defmacro(&self) -> bool {
        matches!(self, Node::Defmacro(_))
    }

    pub fn as_module(&self) -> &Module {
        match self {
            Node::Module(m) => m,
            _ => panic!("expected Module"),
        }
    }

    pub fn as_operation(&self) -> &Operation {
        match self {
            Node::Operation(o) => o,
            _ => panic!("expected Operation"),
        }
    }

    pub fn as_region(&self) -> &Region {
        match self {
            Node::Region(r) => r,
            _ => panic!("expected Region"),
        }
    }

    pub fn as_block(&self) -> &Block {
        match self {
            Node::Block(b) => b,
            _ => panic!("expected Block"),
        }
    }

    pub fn as_def(&self) -> &Binding {
        match self {
            Node::Def(b) => b,
            _ => panic!("expected Def"),
        }
    }

    pub fn as_let(&self) -> &LetExpr {
        match self {
            Node::Let(l) => l,
            _ => panic!("expected Let"),
        }
    }

    pub fn as_type_annotation(&self) -> &TypeAnnotation {
        match self {
            Node::TypeAnnotation(t) => t,
            _ => panic!("expected TypeAnnotation"),
        }
    }

    pub fn as_function_type(&self) -> &FunctionType {
        match self {
            Node::FunctionType(ft) => ft,
            _ => panic!("expected FunctionType"),
        }
    }

    pub fn as_literal(&self) -> &Value {
        match self {
            Node::Literal(v) => v,
            _ => panic!("expected Literal"),
        }
    }

    pub fn as_require(&self) -> &Require {
        match self {
            Node::Require(r) => r,
            _ => panic!("expected Require"),
        }
    }

    pub fn as_require_macros(&self) -> &RequireMacros {
        match self {
            Node::RequireMacros(r) => r,
            _ => panic!("expected RequireMacros"),
        }
    }

    pub fn as_compilation(&self) -> &Compilation {
        match self {
            Node::Compilation(c) => c,
            _ => panic!("expected Compilation"),
        }
    }

    pub fn as_extern(&self) -> &Extern {
        match self {
            Node::Extern(e) => e,
            _ => panic!("expected Extern"),
        }
    }

    pub fn as_defmacro(&self) -> &Defmacro {
        match self {
            Node::Defmacro(d) => d,
            _ => panic!("expected Defmacro"),
        }
    }
}
